tensorboard/plugins/profile/pod_viewer/topology_graph/topology-graph.ts

```diff
@@ -0,0 +1,544 @@
+    let parent = this;
+    const border = 1;
+    const borderColor = 'black';
+    let cards = svg.selectAll('.xdim').data(data, (d) => d.xdim);
+    cards.enter()
+        .append('rect')
+        .attr('id', (d) => 'rid' + d.rid)
+        .attr(
+            'x',
+            (d) => parent.getNodeXLoc(
+                Math.floor(d.xdim / parent.hostXStride_),
+                d.xdim % parent.hostXStride_, d.nid))
+        .attr(
+            'y',
+            (d) => parent.getChipYLoc(
+                Math.floor(d.ydim / parent.hostYStride_),
+                d.ydim % parent.hostYStride_))
+        .attr('rx', 4 / nodesPerChip)
+        .attr('ry', 4)
+        .attr('class', 'hour bordered')
+        .attr('width', parent.nodeGridWidth_)
+        .attr('height', parent.nodeGridHeight_)
+        .attr('border', border)
+        .style('fill', colors[0])
+        .style('stroke', borderColor)
+        .style('stroke-width', border)
+        .merge(cards)
+        .on('mouseover',
+            function(d) {
+              // highlight text
+              d3.select(this).classed('cell-hover', true).style('opacity', 0.5);
+              parent.selectedChipId = d.cid;
+
+              // Update the tooltip position and value
+              d3.select('#tooltip')
+                  .style('left', d3.event.pageX + 10 + 'px')
+                  .style('top', d3.event.pageY - 10 + 'px')
+                  .select('#value')
+                  .text(parent.getToolTipText_(d));
+              d3.select('#tooltip')
+                  .classed('hidden', false);
+            })
+        .on('mouseout',
+            function() {
+              parent.selectedChipId = -1;
+              d3.select(this)
+                  .classed('cell-hover', false)
+                  .style('opacity', 1.0);
+              d3.select('#tooltip').classed('hidden', true);
+            })
+        .transition()
+        .duration(1000)
+        .style('fill', (d) => colorScale(d.value / d.total));
+    cards.exit().remove();
+  },
+  /**
+   * Draw the UI of chip to chip links.
+   */
+  drawLinks: function(linkData) {
+    let parent = this;
+    if (!linkData || linkData.length == 0 || !this.gLink_) {
+      return;
+    }
+
+    // Handle links;
+    let links = this.gLink_.selectAll('.link').data(linkData);
+
+    // attach the arrow from defs
+    links.enter()
+        .append('svg:path')
+        .attr('id', (d) => 'cid' + d.channelId)
+        .attr('stroke-width', 2)
+        .attr('stroke', 'red')
+        .attr('fill', 'none')
+        .attr('marker-end', 'url(#arrow)')
+        .style('visibility', 'hidden')
+        .merge(links)
+        .attr('d', (d) => parent.linkToPath(d));
+
+    // Handle deleted links.
+    links.exit().remove();
+    this.selectedChannelIdChanged_(this.selectedChannelId);
+  },
+  /**
+   * Given the global core id, returns the (x, y) coordinates in the topology
+   * graph.
+   * @return [x, y]
```

Comment:
Instead of this Closure type like notation, I would create a type for this tuple and use it throughout.

Reply:
Done.